Make message links to deleted profiles display 404: (fixes 11020)

jacobguinther opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #11020

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Changes in chatMessages:
  - Move Red 'Deleted Account' Notification Trigger
      - From chatMessages -> profile
  - Other changes in chatMessages affect avatar link

Changes in profile:
  - If user doesnt exist:
    - Render 404
    - Display Red 'Deleted Account' Notification

Now the 'Deleted Account' notification fires on deleted user profile page as well as modals.

Clicking Avatar or Username
![Screen Shot 2020-05-24 at 11 18 18 AM](https://user-images.githubusercontent.com/54478555/82760479-f3e69000-9db0-11ea-861a-b7c37b131c83.png)

Displays 404 modal with notification
![Screen Shot 2020-05-24 at 11 20 14 AM](https://user-images.githubusercontent.com/54478555/82760481-f5b05380-9db0-11ea-8da8-8c55505ee8ef.png)

Deleted Users Profile page now looks like this:
![Screen Shot 2020-05-24 at 11 20 35 AM](https://user-images.githubusercontent.com/54478555/82760492-0791f680-9db1-11ea-89c4-e32372d19eb0.png)
( displays same notification as well (not shown in picture))

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 234319eb-3c97-4cf7-9248-9feff04a31b6

jacobguinther committed:
Make links to deleted profiles display 404

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -836,8 +841,23 @@ export default {
+        } catch (e) {
+          // user = null;
+          user.active = false;
+          console.log('fkdsafj;dsak;', user.active);
```
this should be removed, as well as the other comments like // debugger, //user = null

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -826,7 +831,7 @@ export default {
   methods: {
     async loadUser () {
-      let user = this.userLoggedIn;
+      let user = await this.userLoggedIn;
```
why this await? this.userLoggedIn isn't a promise as far as I know

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -733,6 +737,7 @@ export default {
   components: {
     MemberDetails,
     profileStats,
+    error,
```
can this be renamed to something clearer? like 404Error?

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -836,8 +841,23 @@ export default {
+          user = response.data.data;
+        } catch (e) {
+          // user = null;
+          user.active = false;
```
In case the user cannot be fetched, why not keep `this.user` undefined and in the component above do `v-if="!user"` instead of adding a new property to the user object?

jacobguinther replied:
I agree with you. I don't think I should be adding a new property to the user. However, when I try and do it your suggested way.. I keep getting this bug where the 404 will flash on the modal for every active user before their actual profile displays. Adding the new property is the only way I could get it to work without flashing 404 for active users.

I am still looking into this.

Alys replied:
That seems related to this issue: https://github.com/HabitRPG/habitica/issues/11051
(I.e., it's known that profiles can be slow to load and if caching of a previous profile hasn't happened, you see a blank profile screen before the data loads). That issue is tolerable - we rarely see complaints about it, but if the default state was the 404 error, I'm sure that would cause confusion and more bug reports.

So, I'm a big fan of not flashing up the 404 message initially. :)
If it tunrs out to not be possible to find a way to avoid that without adding a new properly, perhaps we could have the new property as a temporary workaround and then adjust the code later once https://github.com/HabitRPG/habitica/issues/11051 is fixed?

jacobguinther replied:
Sounds good to me. Im having a hard time getting it to work correctly without adding the property. @paglias Any thoughts?

paglias replied:
@jacobguinther can you post the version without the property? I'll give a look and see if I can get it to work

jacobguinther replied:
eh. Found another bug Im working on. 
![logged_in_user_404_bug](https://media.giphy.com/media/S6SICfPVYec0CEdIlJ/giphy.gif)
Currently. If I refresh the page the  logged in users profile modal shows up fine. However if i click any deleted user.. Now the logged in users profile shows 404 also.

I will post version without property once I fix this.

jacobguinther replied:
Ok. I fixed the above bug and got it to work without adding the extra property on user object.

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -836,8 +841,23 @@ export default {
+          });
+          // debugger;
+          user = response.data.data;
+        } catch (e) {
```
This will catch all errors even if they're not 404s

paglias reviewed (changes requested):
left some comments!

jacobguinther committed:
Implement some requested changes

- remove comments
- remove 'await'
- rename error -> 404error
- remove catch block -> just check for 404

jacobguinther committed:
Fix bug: Logged in user profile showing 404

- Initiallty logged in user will display fine
- This bug only appeared after you view a deleted users modal

jacobguinther committed:
Remove 'user.active' attribute

- add change to data():
- user: undefined -> user: {}

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -766,7 +771,7 @@ export default {
       achievementsCategories: {}, // number, open
       content: Content,
-      user: undefined,
+      user: {},
```
why not keep undefined? This way `v-if="!user"` will never evaluate to true because `{}` is considered a truthy value

jacobguinther replied:
This was to keep 404 from flashing on modal. I set user to undefined only if 'fetchMember' returns a 404.

If it wasn't for flashing 404 bug I wouldn't have done this.

jacobguinther replied:
Would you like me to try and find another way to fix the flash 404 bug without changing the initial state of user?

paglias replied:
sorry for the delay, yes if you could try another way it would be great but if it's not possible let me know

jacobguinther replied:
Im having a hard time finding a way around this. Any suggestions?

paglias commented on website/client/src/components/chat/chatMessages.vue:
```diff
@@ -331,6 +326,9 @@ export default {
       if (profile && !profile.rejected) {
```
here (lines 326-331) basically the same thing happens in both cases, if profile is rejected or not because `profile._id` and `memberId` should be the same value

paglias commented on website/client/src/components/chat/chatMessages.vue:
```diff
@@ -315,12 +315,7 @@ export default {
       if (!profile._id) {
```
This is probably unrelated to this PR, but since we're here, we're fetching the profile data (cached or not) and then redirecting to the modal page, so the fetched was not used at all, I guess this whole function could be simplified into 

```
async showMemberModal (memberId) {	
    this.$router.push({ name: 'userProfile', params: { userId: memberId } });
}
```

jacobguinther committed:
Shorten showMemberModal()

- Remove unused fetch profile data
- Remove unnecessary conditional

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -1,8 +1,11 @@
   >
+    <error404 />
+  </div>
+  <div v-else>
```
this is missing the `profile` class

paglias reviewed (changes requested):
Sorry for the delay @jacobguinther but with the following changes all issues seems to be solved:

```diff
diff --git a/website/client/src/components/userMenu/profile.vue b/website/client/src/components/userMenu/profile.vue
index 9a7dc4eb6b..621b7f960b 100644
--- a/website/client/src/components/userMenu/profile.vue
+++ b/website/client/src/components/userMenu/profile.vue
@@ -1,11 +1,10 @@
 <template>
   <div
-    v-if="!user"
-    class="profile"
+    v-if="!user && userLoaded"
   >
     <error404 />
   </div>
-  <div v-else>
+  <div class="profile" v-else-if="userLoaded">
     <div class="header">
       <span
         class="close-icon svg-icon inline icon-10"
@@ -772,7 +771,8 @@ export default {
       achievements: {},
       achievementsCategories: {}, // number, open
       content: Content,
-      user: {},
+      user: null,
+      userLoaded: false,
     };
   },
   computed: {
@@ -832,11 +832,12 @@ export default {
   },
   methods: {
     async loadUser () {
-      let user = {};
+      let user = null;
 
       // Reset editing when user is changed. Move to watch or is this good?
       this.editing = false;
       this.hero = {};
+      this.userLoaded = false;
       this.adminToolsLoaded = false;
 
       const profileUserId = this.userId;
@@ -844,7 +845,7 @@ export default {
       if (profileUserId && profileUserId !== this.userLoggedIn._id) {
         const response = await this.$store.dispatch('members:fetchMember', { memberId: profileUserId });
         if (response.response && response.response.status === 404) {
-          user = undefined;
+          user = null;
           this.$store.dispatch('snackbars:add', {
             title: 'Habitica',
             text: this.$t('messageDeletedUser'),
@@ -880,6 +881,8 @@ export default {
       } else {
         this.user = this.userLoggedIn;
       }
+
+      this.userLoaded = true;
     },
     selectPage (page) {
       this.selectedPage = page || 'profile';
```

jacobguinther committed:
Merge remote-tracking branch 'upstream/develop' into deleted-user-profile--11020

jacobguinther committed:
Change user initial value to null

- Add userLoaded to data

paglias commented:
thanks @jacobguinther ! Noted towards your 2nd tier, sorry it took so long to get this merged
